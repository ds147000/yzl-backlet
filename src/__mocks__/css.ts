export default {
  process(): string {
    return '';
  },
};
