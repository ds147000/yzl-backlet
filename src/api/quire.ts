/* eslint-disable prettier/prettier */
/* eslint-disable import/prefer-default-export */
import request from '@/libs/api';

export type HashBlockDetailsResponse = {
  'hash'?: string;
  'ver'?: number;
  'prev_block'?: string;
  'mrkl_root'?: string;
  'time'?: number;
  'bits'?: number;
  'nonce'?: number;
  'n_tx'?: number;
  'size'?: number;
  'block_index'?: number;
  'main_chain'?: boolean;
  'height'?: number;
  'received_time'?: number;
  'relayed_by'?: string;
  'tx'?: HashBlockDetailsResponse[];
};

export async function GetHashBlockDetails(data: string) {
  return request.get<HashBlockDetailsResponse>(`https://blockchain.info/rawblock/${data}`);
}
