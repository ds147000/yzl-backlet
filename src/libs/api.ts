import axios, { AxiosRequestConfig, AxiosResponse, AxiosStatic, CancelTokenSource } from 'axios';

export class Request {
  private _axios: AxiosStatic;

  constructor(_axios: AxiosStatic) {
    this._axios = _axios;
    this.bindInterceptors();
  }

  bindInterceptors(): void {
    this._axios.interceptors.request.use(this.handleRequest, this.handleRequestError);
    this._axios.interceptors.response.use(this.handleResponse);
  }

  handleRequest(config: AxiosRequestConfig): AxiosRequestConfig {
    const prefix = window.blocklet ? window.blocklet.prefix : '/';
    config.baseURL = prefix || '';
    config.timeout = 200000;
    return config;
  }

  handleResponse(response: AxiosResponse<any>): any {
    return response.data;
  }

  async handleRequestError(error: any): Promise<any> {
    return Promise.reject(error);
  }

  async post<res>(url: string, data?: any): Promise<res> {
    return this._axios.post(url, data);
  }

  async get<res>(url: string, parmas?: any): Promise<res> {
    return this._axios.get(url, parmas);
  }

  getCloseSoure(): CancelTokenSource {
    return axios.CancelToken.source();
  }
}

const request = new Request(axios);

export default request;
