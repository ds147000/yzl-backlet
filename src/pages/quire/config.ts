/* eslint-disable import/prefer-default-export */
/* eslint-disable prettier/prettier */

export const LabelMap = {
  'hash': 'Hash',
  'ver': 'Ver',
  'prev_block': 'PrevBlock',
  'mrkl_root': 'Mrkl Root',
  'time': 'Time',
  'bits': 'Bits',
  'nonce': 'Nonce',
  'n_tx': 'N Text',
  'size': 'Size',
  'block_index': 'Block Index',
  'main_chain': 'Main Chain',
  'height': 'Height',
  'received_time': 'Received Time',
  'relayed_by': 'Relayed By',
  'tx': 'More',
};
