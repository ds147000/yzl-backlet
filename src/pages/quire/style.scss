.quire {
  .header {
    height: unit(100);
    background-color: white;
    box-shadow: 0 0 5px 0 $back;
  }

  .input {
    width: unit(280);
  }

  .list {
    position: relative;
    padding: unit(15) unit(20) unit(40) unit(20);
    box-sizing: border-box;
    width: 100vw;
    overflow: hidden;

    .mask {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      z-index: 5;
      background-color: $mask;
      box-sizing: border-box;
      padding-top: unit(150);
    }

    .item {
      border-bottom: 1px solid $border;
      padding: unit(10) unit(0);

      .label {
        width: unit(115);
        font-weight: 700;
        font-size: unit(14);
        color: $text2;
      }

      .value {
        color: $text;
        font-size: unit(12);
        word-wrap: break-word;
        word-break: break-all;
      }
    }
  }
}

@media screen and (max-width: 750px) {
  .quire {
    .ant-form .ant-form-item .ant-form-item-label {
      display: none;
    }

    .ant-form-inline .ant-form-item {
      margin-right: unit(0);
    }

    .list {
      .mask {
        padding-top: unit(200);
      }

      .item {
        .label {
          font-size: unit(12);
        }

        .value {
          font-size: unit(10);
          width: unit(250);
        }
      }
    }
  }
}
