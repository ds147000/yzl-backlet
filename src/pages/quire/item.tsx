import React from 'react';
import { HashBlockDetailsResponse } from '@/api/quire';
import { LabelMap } from './config';

type ItemValue = string | number | boolean | HashBlockDetailsResponse[];

interface ItemProps {
  label: keyof HashBlockDetailsResponse;
  value?: ItemValue;
}

const Item: React.FC<ItemProps> = ({ label, value }) => {
  return (
    <div className="item flex">
      <div className="label">{LabelMap[label]}</div>
      <div className="flex1 value" data-testid={label}>
        {valueToString(value)}
      </div>
    </div>
  );
};

function valueToString(value?: ItemValue): React.ReactNode {
  if (!value) return '';

  switch (typeof value) {
    case 'boolean':
      return value ? 'Yes' : 'No';

    case 'number':
      return String(value);

    case 'string':
      return value;

    default:
      return value.map((item) => (
        <p key={item.hash}>
          <a href={`/quire?hash=${item.hash}`} target={item.hash}>
            {item.hash}
          </a>
          ;
        </p>
      ));
  }
}

export default Item;
