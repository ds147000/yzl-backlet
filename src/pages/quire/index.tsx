import qs from 'qs';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { Form, Input, Button, message, Empty, Affix, Spin } from 'antd';
import { RouteComponentProps } from 'react-router-dom';
import { GetHashBlockDetails, HashBlockDetailsResponse } from '@/api/quire';
import { LabelMap } from './config';
import Item from './item';
import './style.scss';

const InputRules = [
  {
    required: true,
    message: '请输入 Block Hash',
    transform: (value: string) => String(value).trim(),
  },
];

type RouterPramas = {
  hash?: string;
};

const DelayTime = 100;

const Quire: React.FC<RouteComponentProps> = ({ history }) => {
  const [data, setData] = useState<HashBlockDetailsResponse>({});
  const [loading, setLoading] = useState(false);
  const keys = useMemo(() => Object.keys(LabelMap) as (keyof HashBlockDetailsResponse)[], [data]);
  const isEmpty = Object.values(data).length === 0;
  const params = qs.parse(history.location.search.slice(1)) as RouterPramas;

  const onSubmit = useCallback(
    ({ hash }) => {
      if (params.hash !== hash) history.push(`?hash=${hash}`);

      setLoading(true);
      GetHashBlockDetails(hash)
        .then((res) => setData(res))
        .catch((err) => {
          if (err.response.data?.message) message.error(err.response.data?.message);
        })
        .finally(() => setLoading(false));
    },
    [params.hash]
  );

  useEffect(() => {
    if (params.hash) onSubmit({ hash: params.hash });
  }, []);

  return (
    <div className="page quire">
      <Affix>
        <div className="header flex flex-ac flex-jc">
          <Form layout="inline" onFinish={onSubmit}>
            <Form.Item label="Block Hash" name="hash" rules={InputRules}>
              <Input placeholder="请输入Block Hash" className="input" />
            </Form.Item>
            <Form.Item>
              <Button type="primary" htmlType="submit" block>
                查询
              </Button>
            </Form.Item>
          </Form>
        </div>
      </Affix>
      <Spin spinning={loading} size="large" delay={DelayTime} tip="努力查询中...">
        <div className="list">
          {keys.map((key) => (
            <Item key={key} label={key} value={data[key]} />
          ))}
          {isEmpty && !loading && (
            <div className="mask">
              <Empty description="暂无数据" />
            </div>
          )}
        </div>
      </Spin>
    </div>
  );
};

export default Quire;
