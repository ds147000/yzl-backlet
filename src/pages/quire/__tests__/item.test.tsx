// eslint-disable-next-line no-unused-vars
import React from 'react';
import { render } from '@testing-library/react';
import { GetHashBlockDetailsReuslt } from '@/api/__mocks__/quire';
import { HashBlockDetailsResponse } from '@/api/quire';
import Item from '../item';

describe('Quire/Item', () => {
  test('boolean值, false', () => {
    const Page = render(<Item label="bits" value={false} />);
    expect(Page.getByTestId('bits').innerHTML).toBe('No');
  });

  test('boolean值, true', () => {
    const Page = render(<Item label="bits" value />);
    expect(Page.getByTestId('bits').innerHTML).toBe('Yes');
  });

  test('多个值数组, true', () => {
    const newGetHashBlockDetailsReuslt = { ...GetHashBlockDetailsReuslt };
    newGetHashBlockDetailsReuslt.hash = '12345678';
    const Page = render(<Item label="bits" value={[GetHashBlockDetailsReuslt, newGetHashBlockDetailsReuslt]} />);
    const A1 = Page.getByText(GetHashBlockDetailsReuslt.hash) as HTMLAnchorElement;
    const A2 = Page.getByText(newGetHashBlockDetailsReuslt.hash) as HTMLAnchorElement;
    expect(A1.href).toBe(`http://localhost/quire?hash=${GetHashBlockDetailsReuslt.hash}`);
    expect(A2.href).toBe(`http://localhost/quire?hash=${newGetHashBlockDetailsReuslt.hash}`);
  });

  test('快照', () => {
    const Page = render(
      <div>
        {Object.keys(GetHashBlockDetailsReuslt).map((key) => (
          <Item
            key={key}
            label={key as keyof HashBlockDetailsResponse}
            value={GetHashBlockDetailsReuslt[key as keyof HashBlockDetailsResponse]}
          />
        ))}
      </div>
    );
    expect(Page).toMatchSnapshot();
  });
});
