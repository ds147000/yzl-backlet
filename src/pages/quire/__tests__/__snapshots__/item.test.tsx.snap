// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quire/Item 快照 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Hash
          </div>
          <div
            class="flex1 value"
            data-testid="hash"
          >
            0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Ver
          </div>
          <div
            class="flex1 value"
            data-testid="ver"
          >
            1
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            PrevBlock
          </div>
          <div
            class="flex1 value"
            data-testid="prev_block"
          >
            00000000000007d0f98d9edca880a6c124e25095712df8952e0439ac7409738a
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Mrkl Root
          </div>
          <div
            class="flex1 value"
            data-testid="mrkl_root"
          >
            935aa0ed2e29a4b81e0c995c39e06995ecce7ddbebb26ed32d550a72e8200bf5
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Time
          </div>
          <div
            class="flex1 value"
            data-testid="time"
          >
            1322131230
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Bits
          </div>
          <div
            class="flex1 value"
            data-testid="bits"
          >
            437129626
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Nonce
          </div>
          <div
            class="flex1 value"
            data-testid="nonce"
          >
            2964215930
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            N Text
          </div>
          <div
            class="flex1 value"
            data-testid="n_tx"
          >
            22
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Size
          </div>
          <div
            class="flex1 value"
            data-testid="size"
          >
            9195
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Block Index
          </div>
          <div
            class="flex1 value"
            data-testid="block_index"
          >
            818044
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Main Chain
          </div>
          <div
            class="flex1 value"
            data-testid="main_chain"
          >
            Yes
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Height
          </div>
          <div
            class="flex1 value"
            data-testid="height"
          >
            154595
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Received Time
          </div>
          <div
            class="flex1 value"
            data-testid="received_time"
          >
            1322131301
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            Relayed By
          </div>
          <div
            class="flex1 value"
            data-testid="relayed_by"
          >
            108.60.208.156
          </div>
        </div>
        <div
          class="item flex"
        >
          <div
            class="label"
          >
            More
          </div>
          <div
            class="flex1 value"
            data-testid="tx"
          >
            <p>
              <a
                href="/quire?hash=0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103"
                target="0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103"
              >
                0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103
              </a>
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Hash
        </div>
        <div
          class="flex1 value"
          data-testid="hash"
        >
          0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Ver
        </div>
        <div
          class="flex1 value"
          data-testid="ver"
        >
          1
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          PrevBlock
        </div>
        <div
          class="flex1 value"
          data-testid="prev_block"
        >
          00000000000007d0f98d9edca880a6c124e25095712df8952e0439ac7409738a
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Mrkl Root
        </div>
        <div
          class="flex1 value"
          data-testid="mrkl_root"
        >
          935aa0ed2e29a4b81e0c995c39e06995ecce7ddbebb26ed32d550a72e8200bf5
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Time
        </div>
        <div
          class="flex1 value"
          data-testid="time"
        >
          1322131230
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Bits
        </div>
        <div
          class="flex1 value"
          data-testid="bits"
        >
          437129626
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Nonce
        </div>
        <div
          class="flex1 value"
          data-testid="nonce"
        >
          2964215930
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          N Text
        </div>
        <div
          class="flex1 value"
          data-testid="n_tx"
        >
          22
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Size
        </div>
        <div
          class="flex1 value"
          data-testid="size"
        >
          9195
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Block Index
        </div>
        <div
          class="flex1 value"
          data-testid="block_index"
        >
          818044
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Main Chain
        </div>
        <div
          class="flex1 value"
          data-testid="main_chain"
        >
          Yes
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Height
        </div>
        <div
          class="flex1 value"
          data-testid="height"
        >
          154595
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Received Time
        </div>
        <div
          class="flex1 value"
          data-testid="received_time"
        >
          1322131301
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          Relayed By
        </div>
        <div
          class="flex1 value"
          data-testid="relayed_by"
        >
          108.60.208.156
        </div>
      </div>
      <div
        class="item flex"
      >
        <div
          class="label"
        >
          More
        </div>
        <div
          class="flex1 value"
          data-testid="tx"
        >
          <p>
            <a
              href="/quire?hash=0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103"
              target="0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103"
            >
              0000000000000bae09a7a393a8acded75aa67e46cb81f7acaa5ad94f9eacd103
            </a>
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
